internal/controller/garbagecollector_controller.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	nacv1alpha1 "github.com/migtools/oadp-non-admin/api/v1alpha1"
+	"github.com/migtools/oadp-non-admin/internal/common/constant"
+	"github.com/migtools/oadp-non-admin/internal/common/function"
+	"github.com/migtools/oadp-non-admin/internal/source"
+)
+
+// GarbageCollectorReconciler reconciles Velero objects
+type GarbageCollectorReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	OADPNamespace string
+	Frequency     time.Duration
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *GarbageCollectorReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	labelSelector := client.MatchingLabels{
+		constant.OadpLabel:      constant.OadpLabelValue,
+		constant.ManagedByLabel: constant.ManagedByLabelValue,
+	}
+
+	logger.V(1).Info("Garbage Collector Reconcile start")
+	// TODO duplication in delete logic
+	// TODO do deletion in parallel?
+
+	secretList := &corev1.SecretList{}
+	if err := r.List(ctx, secretList, client.InNamespace(r.OADPNamespace), labelSelector); err != nil {
+		logger.Error(err, "Unable to fetch Secret in OADP namespace")
+		return ctrl.Result{}, err
+	}
+	for _, secret := range secretList.Items {
+		annotations := secret.GetAnnotations()
+		// TODO check UUID as well?
```

Comment:
As we are using UUID to establish relationship among objects, lets use it to confirm whether we want to classify the object as orphan/residual or not before deleting it.

Reply:
done